backends/p4tools/testgen/core/small_step/table_stepper.cpp

```diff
@@ -0,0 +1,662 @@
+                    auto* clonedTableProperties = p4Table->properties->clone();
+                    auto* properties = &clonedTableProperties->properties;
+                    auto* propertyValue = properties->at(propertyIdx)->clone();
+                    const auto* key = clonedTable->getKey();
+                    CHECK_NULL(key);
+                    auto* newKey = key->clone();
+                    auto* newKeyElement = newKey->keyElements[keyIdx]->clone();
+                    // Now bubble them up in reverse order.
+                    // Replace the single (!) key in the table.
+                    newKeyElement->expression = v->param;
+                    newKey->keyElements[keyIdx] = newKeyElement;
+                    propertyValue->value = newKey;
+                    (*properties)[propertyIdx] = propertyValue;
+                    clonedTable->properties = clonedTableProperties;
+                    return Continuation::Return(clonedTable);
+                });
+            return true;
+        }
+
+        const auto* nameAnnot = keyElement->getAnnotation("name");
+        // Some hidden tables do not have any key name annotations.
+        BUG_CHECK(nameAnnot != nullptr || properties.tableIsImmutable,
+                  "Non-constant table key without an annotation");
+        cstring fieldName;
+        if (nameAnnot != nullptr) {
+            fieldName = nameAnnot->getName();
+        }
+        // It is actually possible to use a variety of types as key.
+        // So we have to stay generic and produce a corresponding variable.
+        cstring keyMatchType = keyElement->matchType->toString();
+        // We can recover from taint for some match types, which is why we track taint.
+        bool keyHasTaint = state->hasTaint(keyElement->expression);
+
+        // If we match with a boolean value, replace the value with a mux that returns 0 or 1
+        // depending on the value of the boolean key.
+        // This expression is symbolic, as key->expression has already been resolved at this
+        // point.
+        if (keyElement->expression->type->is<IR::Type_Boolean>()) {
+            auto* clonedKey = keyElement->clone();
+            const auto* oneBitType = IRUtils::getBitType(1);
+            const auto* falseBit = IRUtils::getConstant(oneBitType, 0);
+            const auto* trueBit = IRUtils::getConstant(oneBitType, 1);
+            clonedKey->expression =
+                new IR::Mux(oneBitType, keyElement->expression, trueBit, falseBit);
+            keyElement = clonedKey;
+        }
+
+        // Initialize the standard keyProperties.
+        KeyProperties keyProperties(keyElement, fieldName, keyIdx, keyMatchType, keyHasTaint);
+        properties.resolvedKeys.emplace_back(keyProperties);
+    }
+    return false;
+}
+
+bool TableStepper::checkTableIsImmutable() {
+    bool isConstant = false;
+    if (table->properties->getProperty("entries") != nullptr) {
```

Comment:
there is a getEntries method

Reply:
This method returns and EntriesList, which does not have the `isConstant` member.